### DISCUSSION TITLE
What is the recommended way to save particle temperatures?

### DISCUSSION THREAD

**Discussion by ghost** [BLAST-WarpX/warpx#4296] [Simulation Inputs]
Is there an easy way to save particle temperatures? I'm having trouble trying to put temperature into the reduced diags. Is it not as straightforward as it is with ParticleNumber, ParticleEnergy, and ParticleMomentum? Or do I need to calculate the value?

Thanks!

**Comment by ax3l** [comment]
Thanks for the great question @Nsh002!

A few details before we start: WarpX is a kinetic modeling code and as such, our particles are most of the time non-Maxwellian. Thus, for a general simulation, we can guide you how to calculate moments or distributions of your particles in a simulation - but you need to check they are actually resembling a "temperature" for your scenario of interest.

In output, we can output particles and fields. If you output particles, then you can calculate distributions and moments in post-processing from the individual macro-particles.

Alternatively, you can also in situ derive field values from the particles per grid cell. The [WarpX full diagnostics](https://warpx.readthedocs.io/en/latest/usage/parameters.html#full-diagnostics) have these options:
- `<diag_name>.particle_fields_to_plot`
- `<diag_name>.particle_fields.<field_name>(x,y,z,ux,uy,uz)`

That you can use to calculate an average momentum (or average kinetic energy) from the `ux`, `uy`, `uz` properties of individual particles. For example, the average momentum would be:
```
diagnostics.diags_names = diag1
diag1.intervals = 200
diag1.diag_type = Full
diag1.fields_to_plot = Ex Ey Ez Bx By Bz jx jy jz rho rho_electrons

diag1.particle_fields_to_plot = avg_e_mom
diag1.particle_fields_species = electrons
diag1.particle_fields.avg_e_mom(x,y,z,ux,uy,uz) = sqrt(ux*ux + uy*uy + uz*uz)
diag1.particle_fields.avg_e_mom.do_average = 0
```
